llvm/test/CodeGen/Xtensa/float-fma.ll

```diff
@@ -0,0 +1,147 @@
+  %negadd = fneg float %add
+  ret float %negadd
+}
+
+
+define float @fnmsub_s(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fnmsub_s:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    msub.s f10, f9, f8
+; XTENSA-NEXT:    rfr a2, f10
+; XTENSA-NEXT:    retw.n
+  %nega = fneg float %a
+  %mul = fmul float %nega, %b
+  %add = fadd float %mul, %c
+  ret float %add
+}
+
+declare float @llvm.fma.f32(float, float, float)
+
+define float @fmadd_s_intrinsics(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fmadd_s_intrinsics:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    madd.s f10, f9, f8
+; XTENSA-NEXT:    rfr a2, f10
+; XTENSA-NEXT:    retw.n
+  %fma = call float @llvm.fma.f32(float %a, float %b, float %c)
+  ret float %fma
+}
+
+define float @fmsub_s_intrinsics(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fmsub_s_intrinsics:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    neg.s f10, f10
+; XTENSA-NEXT:    madd.s f10, f9, f8
+; XTENSA-NEXT:    rfr a2, f10
+; XTENSA-NEXT:    retw.n
+  %negc = fneg float %c
+  %fma = call float @llvm.fma.f32(float %a, float %b, float %negc)
+  ret float %fma
+}
+
+define float @fnmadd_s_intrinsics(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fnmadd_s_intrinsics:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    madd.s f10, f9, f8
+; XTENSA-NEXT:    neg.s f8, f10
+; XTENSA-NEXT:    rfr a2, f8
+; XTENSA-NEXT:    retw.n
+  %fma = call float @llvm.fma.f32(float %a, float %b, float %c)
+  %neg = fneg float %fma
+  ret float %neg
+}
+
+define float @fnmsub_s_intrinsics(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fnmsub_s_intrinsics:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    msub.s f10, f9, f8
+; XTENSA-NEXT:    rfr a2, f10
+; XTENSA-NEXT:    retw.n
+  %nega = fneg float %a
+  %fma = call float @llvm.fma.f32(float %nega, float %b, float %c)
+  ret float %fma
+}
+
+define float @fnmsub_s_swap_intrinsics(float %a, float %b, float %c) nounwind {
+; XTENSA-LABEL: fnmsub_s_swap_intrinsics:
+; XTENSA:       # %bb.0:
+; XTENSA-NEXT:    entry a1, 32
+; XTENSA-NEXT:    wfr f8, a3
+; XTENSA-NEXT:    wfr f9, a2
+; XTENSA-NEXT:    wfr f10, a4
+; XTENSA-NEXT:    neg.s f10, f10
+; XTENSA-NEXT:    madd.s f10, f9, f8
+; XTENSA-NEXT:    rfr a2, f10
+; XTENSA-NEXT:    retw.n
+  %negc = fneg float %c
+  %fma = call float @llvm.fma.f32(float %a, float %b, float %negc)
```

Comment:
@zRedShift , I think all is good and this PR is ready to merge. One question about "fnmsub_s_swap_intrinsics". It looks the same as "fmsub_s_intrinsics" intrinsic. Or I missed something?

Reply:
@andreisfr just saw this in the email, and you're right. I copied these from LoongArch/float-fma.ll, and it looks like fnmsub_s_swap_intrinsics was supposed to be 
```llvm
  %negb = fneg float %b
  %fma = call float @llvm.fma.f32(float %a, float %negb, float %c)
  ret float %fma
```
I'll fix this

Reply:
@zRedShift , I combined  the third commit with the second commit from this PR to single commit and merged all commits with main branch. Thank you very much!